sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/XmlExternalEntityShouldNotBeParsed_XmlReader_Net452.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System.Xml;
+
+namespace SonarAnalyzer.UnitTest.TestCases
+{
+    public class XmlReaderUnsafe
```

Comment:
in [Microsoft's guide](https://docs.microsoft.com/en-us/archive/msdn-magazine/2009/november/xml-denial-of-service-attacks-and-defenses#defending-against-external-entity-attacks) which is referenced in the [OWASP guide](https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html#net), they tell how to enable safe XXE processing

search for "Figure 6 Applying Safer Entity Parsing Settings to XmlDocument"

```
void processXml(string xml)
{
    MemoryStream stream =
        new MemoryStream(Encoding.Default.GetBytes(xml));
    XmlReaderSettings settings = new XmlReaderSettings();

    // allow entity parsing but do so more safely
    settings.ProhibitDtd = false;
    settings.MaxCharactersFromEntities = 1024;
    settings.XmlResolver = new XmlSafeResolver();

    XmlReader reader = XmlReader.Create(stream, settings);
    XmlDocument doc = new XmlDocument();
    doc.Load(reader);
}
```
*if possible* it would be cool to avoid this known FP

Reply:
We don't have a false positive here since we check the exact type of the XmlResolver, similar with what has been done for the other XXE rule types.

The blacklist contains only XmlUrlResolver and XmlPreloadedResolver for now.

I've added a test case though to document the behavior.